src/main/java/appeng/util/helpers/ItemComparisonHelper.java

```diff
@@ -0,0 +1,443 @@
+		 * if ( Mode != FuzzyMode.IGNORE_ALL ) { if ( a.hasTagCompound() && !isShared( a.getTagCompound() ) ) { a =
+		 * Platform.getSharedItemStack( AEItemStack.create( a ) ); }
+		 * if ( b.hasTagCompound() && !isShared( b.getTagCompound() ) ) { b = Platform.getSharedItemStack(
+		 * AEItemStack.create( b ) ); }
+		 * // test regular items with damage values and what not... if ( isShared( a.getTagCompound() ) && isShared(
+		 * b.getTagCompound() ) && a.itemID == b.itemID ) { return ((AppEngSharedNBTTagCompound)
+		 * a.getTagCompound()).compareFuzzyWithRegistry( (AppEngSharedNBTTagCompound) b.getTagCompound() ); } }
+		 */
+
+		return a.isItemEqual( b );
+	}
+
+	/**
+	 * recursive test for NBT Equality, this was faster then trying to compare / generate hashes, its also more reliable
+	 * then the vanilla version which likes to fail when NBT Compound data changes order, it is pretty expensive
+	 * performance wise, so try an use shared tag compounds as long as the system remains in AE.
+	 */
+	public boolean isNbtTagEqual( final NBTBase left, final NBTBase right )
+	{
+		// same type?
+		final byte id = left.getId();
+		if( id == right.getId() )
+		{
+			switch( id )
+			{
+				case 10:
+				{
+					final NBTTagCompound ctA = (NBTTagCompound) left;
+					final NBTTagCompound ctB = (NBTTagCompound) right;
+
+					final Set<String> cA = ctA.getKeySet();
+					final Set<String> cB = ctB.getKeySet();
+
+					if( cA.size() != cB.size() )
+					{
+						return false;
+					}
+
+					for( final String name : cA )
+					{
+						final NBTBase tag = ctA.getTag( name );
+						final NBTBase aTag = ctB.getTag( name );
+						if( aTag == null )
+						{
+							return false;
+						}
+
+						if( !isNbtTagEqual( tag, aTag ) )
+						{
+							return false;
+						}
+					}
+
+					return true;
+				}
+
+				case 9: // ) // A instanceof NBTTagList )
+				{
+					final NBTTagList lA = (NBTTagList) left;
+					final NBTTagList lB = (NBTTagList) right;
+					if( lA.tagCount() != lB.tagCount() )
+					{
+						return false;
+					}
+
+					final List<NBTBase> tag = tagList( lA );
+					final List<NBTBase> aTag = tagList( lB );
+					if( tag.size() != aTag.size() )
+					{
+						return false;
+					}
+
+					for( int x = 0; x < tag.size(); x++ )
+					{
+						if( aTag.get( x ) == null )
+						{
+							return false;
+						}
+
+						if( !isNbtTagEqual( tag.get( x ), aTag.get( x ) ) )
+						{
+							return false;
+						}
+					}
+
+					return true;
+				}
+
+				case 1: // ( A instanceof NBTTagByte )
+					return ( (NBTPrimitive) left ).getByte() == ( (NBTPrimitive) right ).getByte();
+
+				case 4: // else if ( A instanceof NBTTagLong )
+					return ( (NBTPrimitive) left ).getLong() == ( (NBTPrimitive) right ).getLong();
+
+				case 8: // else if ( A instanceof NBTTagString )
+					return ( (NBTTagString) left ).getString().equals( ( (NBTTagString) right ).getString() ) || ( (NBTTagString) left ).getString().equals( ( (NBTTagString) right ).getString() );
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Identical sub-expressions on both sides of operator "||" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1764)